localstack-core/localstack/services/events/event_rule_engine.py

```diff
@@ -0,0 +1,566 @@
+                    if isinstance(values, dict):
+                        # If the current key has child dict -- key: "key1", child: {"key2": ["val1", val2"]}
+                        # We only update the parent_key and traverse its children with the current branches
+                        array = _traverse_event_pattern(values, array, _parent_key)
+                    else:
+                        # If the current key has no child, this means we found the values to match -- child: ["val1", val2"]
+                        # we update the branches with the parent chain and the values -- {"key1.key2": ["val1, val2"]}
+                        array = [{**item, _parent_key: values} for item in array]
+
+            return array
+
+        return _traverse_event_pattern(nested_dict)
+
+    @staticmethod
+    def flatten_payload(nested_dict: dict) -> list[dict]:
+        """
+        Takes a dictionary as input and will output the dictionary on a single level.
+        The dictionary can have lists containing other dictionaries, and one root level entry will be created for every
+        item in a list.
+        Input:
+        `{"field1": {
+            "field2: [
+                {"field3: "val1", "field4": "val2"},
+                {"field3: "val3", "field4": "val4"},
+            }
+        ]}`
+        Output:
+        `[
+            {
+                "field1.field2.field3": "val1",
+                "field1.field2.field4": "val2"
+            },
+            {
+                "field1.field2.field3": "val3",
+                "field1.field2.field4": "val4"
+            },
+        ]`
+        :param nested_dict: a (nested) dictionary
+        :return: flatten_dict: a dictionary with no nested dict inside, flattened to a single level
+        """
+
+        def _traverse(_object: dict, array=None, parent_key=None) -> list:
+            if isinstance(_object, dict):
+                for key, values in _object.items():
+                    # We update the parent key do that {"key1": {"key2": ""}} becomes "key1.key2"
+                    _parent_key = f"{parent_key}.{key}" if parent_key else key
+                    array = _traverse(values, array, _parent_key)
+
+            elif isinstance(_object, list):
+                if not _object:
+                    return array
+                array = [i for value in _object for i in _traverse(value, array, parent_key)]
+            else:
+                array = [{**item, parent_key: _object} for item in array]
+            return array
+
+        return _traverse(nested_dict, array=[{}], parent_key=None)
+
+
+class EventPatternValidator:
+    def __init__(self):
+        self.error_prefix = "Event pattern is not valid. Reason: "
+
+    def validate_event_pattern(self, event_pattern: str | dict) -> dict[str, t.Any]:
+        if isinstance(event_pattern, str):
+            try:
+                event_pattern = json.loads(event_pattern)
```

Comment:
Do we have a test for that? I don't think the string-based matching is currently implemented in this engine at all? it is expecting `dict` all the way. Could you link me to somewhere I could validate this? 

But yes a plain string is a valid JSON, however it will fail anyway because we are expecting and validating the event is a dict. I believe in Pipes, it should not use this engine if both body and pattern are string, this should only be used to evaluate dicts. 

Happy to jump on a call to discuss this

Reply:
after discussing this on a call, we came to the conclusion that what AWS is calling "plain string" is actually the `body` (or sometimes another field like `data` for Kinesis). You can provider valid JSON inside the body of your SQS message, and ESM/Pipes will automatically decode it for you before passing it down to the EventRuleEngine (or `event-ruler` for AWS), so that you can access nested fields.

This should be part of ESM/Pipes pollers (as it's different for every service) and not the responsibility of the engine, so this is resolved! ✅ thanks a lot @joe4dev for the quick call, we got a lot of great insights! 🙏